File: src/Day14.c
#include "Day14.h"

#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

#include "parsing.h"
#include "hashmap.h"

typedef enum {
  MEM_MASK,
  MEM_SET,
} MemCommandType;

typedef struct {
  MemCommandType type;
  union {
    uint64_t and_mask;
    uint64_t address;
  };
  union {
    uint64_t or_mask;
    uint64_t value;
  };
} MemCommand;

void print_mask(uint64_t num) {
  for (int i = 35; i >= 0; i--) {
    printf("%d", (num >> i) & 1);
  }
  printf("\n");
}

static MemCommand parse_mask(const char* line) {
  MemCommand command = {0};
  command.type = MEM_MASK;
  
  for (int i = 7; line[i]; i++) {
    switch (line[i]) {
      case 'X': continue;
      case '0': command.and_mask |= 1 << (35 - i + 7); break;
      case '1': command.or_mask |= 1 << (35 - i + 7); break;
      case '\n': break;
      default: printf("Bad digit '%c'\n", line[i]); exit(EXIT_FAILURE);
    }
  }
  command.and_mask = (~command.and_mask) & 0xFFFFFFFFF;
  return command;
}

static MemCommand parse_set(const char* line) {
  MemCommand command = {0};
  command.type = MEM_SET;

  sscanf(line, "mem[%llu] = %llu", &command.address, &command.value);
  return command;
}

MemCommand* parse(const char* filename, int* count) {
  FILE* fp;
  fp = fopen(filename, "r");
  if (fp == NULL) {
    printf("Couldn't open input file.\n");
    exit(EXIT_FAILURE);
  }

  *count = count_lines(fp);
  MemCommand* commands = (MemCommand*)malloc(sizeof(MemCommand) * *count);

  for (int i = 0; i < *count; i++) {
    char line[100] = {0};
    fgets(line, 100, fp);
    if (line[1] == 'a') commands[i] = parse_mask(line);
    else if (line[1] == 'e') commands[i] = parse_set(line);
    else {
      printf("Bad line: '%s'\n", line);
      exit(EXIT_FAILURE);
    }
  }

  fclose(fp);
  return commands;
}

uint64_t part1(const char* filename) {
  int count = 0;
  MemCommand* commands = parse(filename, &count);

  HashMap* memory = hashmap_create();
  MemCommand mask;

  for (int i = 0; i < count; i++) {
    MemCommand c = commands[i];
    switch (c.type) {
      case MEM_MASK: mask = c; break;
      case MEM_SET: {
        uint64_t new_value = (c.value & mask.and_mask) | mask.or_mask;
        bool result = hashmap_set(memory, c.address, new_value);
        if (!result) {
          printf("Couldn't set.\n");
          exit(EXIT_FAILURE);
        }
      }
    }
  }
  
  uint64_t total = 0;
  for (int i = 0; i < memory->capacity; i++) {
    if (memory->entries[i].taken) total += memory->entries[i].value;
  }
  hashmap_free(memory);
  free(commands);
  return total;
}

int part2(const char* filename) {
  
}

int day14() {
  printf("====== Day 14 ======\n");
  printf("Part 1: %llu\n", part1("data/day14.txt"));
  printf("Part 2: %d\n", part2("data/day14.txt"));
  return EXIT_SUCCESS;
}

File: src/Day14.py
"""Day 14: Docking Data

Interface with a docking computer doing bit masking to store values
in memory.
"""

from itertools import product
from typing import Generator, Union, List

class MaskCommand:
    """A command that parses and sets the mask for subsequent memory
    setting commands to be filtered through.

    Properties:
        bits (list(int)): The original bit-string of 1's, 0's, and X's.
        and_ (int): A value to be `&`-ed with memory values to set some
                    bits to 0 no matter what
        or_ (int): A value to be `|`-ed with memory values to set some
                    bits to 1 no matter what
        floating (list(int)): Indices where X's live, to help with
                              generating floating bits.
    """
    def __init__(self, s: str):
        self.bits = s
        self.and_ = 0
        self.or_ = 0
        self.floating: list(int) = []
        for i, c in enumerate(s):
            if c == '0':
                self.and_ |= (1 << (35 - i))
            elif c == '1':
                self.or_ |= (1 << (35 - i))
            elif c == 'X':
                self.floating.append(i)
        self.and_ = (~self.and_) & 0xFFFFFFFFF

    @staticmethod
    def generate_binary(digits: int) -> Generator[str, None, None]:
        """Generates strings of binary digits for all values between
        0 and 2**(digits)-1 (i.e. all possible sequences of 1's and 0's.)
        """
        for i in range(1 << digits):
            yield f"{i:0{digits}b}"

    def v2_addresses(self, command: Union['MaskCommand', 'SetCommand']) -> Generator[List[str], None, None]:
        """Generates addresses following V2 guidelines.
        
        Performs an OR between the command's address and the mask, and
        then generates every possible string made up of 1's and 0's in
        place of all the floating bits (X's).

        The OR assumes X's take precedence.

        e.g. 110X0X (after the OR) would yield 110000, 110001, 110100,
        and 110101.
        """ 
        merged = [c_bit if m_bit == '0' else m_bit
                  for (m_bit, c_bit) in zip(self.bits, command.bits)]
        for comb in MaskCommand.generate_binary(len(self.floating)):
            for index, bit in zip(self.floating, comb):
                merged[index] = bit
            yield merged


class SetCommand:
    """A command to set memory to some value.  Stores the address and
    value to be stored there.  Used for both V1 and V2, so includes both
    and integer version of the address and the bits for compatability.

    Properties:
        value (int): The value to be stored.
        address (int): The location in memory to store the value.
        bits (str): A string representation of the 0's and 1's that
                    make up the binary value for the address.  (36 bits wide).
    """
    def __init__(self, address: str, value: str):
        self.value = int(value)
        self.address = int(address)
        self.bits = f"{self.address:036b}"


def parse(filename: str) -> List[Union[MaskCommand, SetCommand]]:
    """Parse the input file into a list of commands."""
    commands = []
    with open(filename, "r") as f:
        for line in f:
            if line.startswith("mask"):
                commands.append(MaskCommand(line[7:]))
            else:
                parts = line.split()
                commands.append(SetCommand(parts[0][4:-1], parts[-1]))
    return commands


def part1(commands: List[Union[MaskCommand, SetCommand]]) -> int:
    """Find the sum of all values in memory after running all commands
    following the V1 spec.
    """
    memory = dict()
    mask = None
    for command in commands:
        if isinstance(command, MaskCommand):
            mask = command
        else:
            new_value = (command.value & mask.and_) | mask.or_
            memory[command.address] = new_value
    
    return sum(memory.values())


def part2(commands: List[Union[MaskCommand, SetCommand]]) -> int:
    """Find the sum of all values in memory after running all commands
    following the V2 spec.
    """
    memory = dict()
    mask = None
    for command in commands:
        if isinstance(command, MaskCommand):
            mask = command
        else:
            for address in mask.v2_addresses(command):
                memory["".join(address)] = command.value
    return sum(memory.values())


if __name__ == "__main__":
    assert 165 == part1(parse("data/day14_test.txt"))
    assert 208 == part2(parse("data/day14_test2.txt"))
    print("All tests passed.")
    commands = parse("data/day14.txt")
    print(part1(commands))
    print(part2(commands))